presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java

```diff
@@ -446,12 +479,38 @@ public static long toPrestoTimestamp(LocalDateTime localDateTime)
     @Deprecated
     public static LocalDateTime fromPrestoLegacyTimestamp(long value, ZoneId sessionZone)
     {
-        return Instant.ofEpochMilli(value).atZone(sessionZone).toLocalDateTime();
+        return fromPrestoLegacyTimestamp(createTimestampType(3), value, sessionZone);
+    }
+
+    /**
+     * @deprecated applicable in legacy timestamp semantics only
+     */
+    @Deprecated
+    public static LocalDateTime fromPrestoLegacyTimestamp(TimestampType timestampType, long value, ZoneId sessionZone)
+    {
+        return getInstantFromTimestamp(timestampType, value).atZone(sessionZone).toLocalDateTime();
     }
 
     public static LocalDateTime fromPrestoTimestamp(long value)
     {
-        return Instant.ofEpochMilli(value).atZone(UTC).toLocalDateTime();
+        return fromPrestoTimestamp(createTimestampType(3), value);
+    }
+
+    public static LocalDateTime fromPrestoTimestamp(TimestampType timestampType, long value)
+    {
+        return getInstantFromTimestamp(timestampType, value).atZone(UTC).toLocalDateTime();
+    }
+
+    private static Instant getInstantFromTimestamp(TimestampType timestampType, long value)
+    {
+        int precision = timestampType.getPrecision();
+        if (precision >= 0 && precision <= 3) {
+            return Instant.ofEpochMilli(value);
+        }
+        if (precision > 3 && precision <= 6) {
+            return Instant.ofEpochSecond(value / 1_000_000, value % 1_000_000 * 1000);
+        }
+        throw new IllegalArgumentException("Unsupported timestamp precision: " + precision);
```

Comment:
If this goes to the user, it should be `PrestoException(NOT_SUPPORTED)`

Can you add a test that would hit this?

Reply:
This can actually never happen if we're dealing with a "short" timestamp (i.e., precision < 6, `long` representation). So it's fine to make it an IllegalArgumentException.